atspi-connection/src/p2p.rs

```diff
@@ -0,0 +1,746 @@
+
+		for app in accessible_applications {
+			let accessible_proxy = app.as_accessible_proxy(conn).await?;
+			let proxies = accessible_proxy.proxies().await?;
+			let application_proxy = proxies.application().await?;
+
+			// Get the application bus address
+			// aka: Does the application support P2P connections?
+			if let Ok(address) = application_proxy.get_application_bus_address().await {
+				let bus_name = BusName::from(app.name);
+				let peer = Peer::try_new(bus_name, address.as_str(), conn).await?;
+				peers.push(peer);
+			}
+		}
+
+		Ok(Peers { peers: Arc::new(Mutex::new(peers)) })
+	}
+
+	/// Returns a [`Peer`] by its bus name.
+	///
+	/// # Note
+	/// Bus names are initialized from `ObjectRef` names, which are `OwnedUniqueName`s.
+	/// This means that the bus name should be a unique name, not a well-known name.
+	///
+	/// # Examples
+	/// ```rust
+	/// # use tokio;
+	/// # use atspi::AccessibilityConnection;
+	/// # use atspi::BusName;
+	/// # use atspi::AtspiResult;
+	/// # use atspi-connection::p2p::Peer;
+	/// #[tokio::main]
+	/// async fn main() -> AtspiResult<()> {
+	/// let conn = AccessibilityConnection::new()?;
+	/// let bus_name = BusName::from(":1.4242");
+	/// let peer: Option<Peer> = conn.get_peer(&bus_name).await;
+	/// # assert!(peer.is_none(), "Peer should not be found");
+	/// # Ok::<(), AtspiResult>(())
+	/// }
+	/// ```
+	pub async fn find_peer(&self, bus_name: &BusName<'_>) -> Option<Peer> {
+		let peers = self.peers.lock().await;
+
+		let matched = match bus_name {
+			BusName::Unique(unique_name) => {
+				peers.iter().find(|peer| peer.unique_name() == unique_name)
+			}
+			BusName::WellKnown(well_known_name) => {
+				let owned_well_known_name = OwnedWellKnownName::from(well_known_name.clone());
+				peers
+					.iter()
+					.find(|peer| peer.well_known_name() == Some(&owned_well_known_name))
+			}
+		};
+		matched.cloned()
+	}
+
+	/// Returns the inner `Arc<Mutex<Vec<Peer>>>`.
+	pub(crate) fn inner(&self) -> Arc<Mutex<Vec<Peer>>> {
+		self.peers.clone()
+	}
+
+	/// Inserts a new `Peer` into the list of peers.
+	async fn insert_unique(
+		&self,
+		unique_name: OwnedUniqueName,
+		conn: &zbus::Connection,
+	) -> AtspiResult<()> {
+		let bus_name = BusName::Unique(unique_name.as_ref());
+		let address = bus_name.get_p2p_address(conn).await?;
+		let p2p_connection = Builder::address(address.clone())?.p2p().build().await?;
+
+		let peer =
+			Peer { unique_name, well_known_name: None, socket_address: address, p2p_connection };
+
+		self.peers.lock().await.push(peer);
+		Ok(())
+	}
+
+	/// Removes a `Peer` from the list of peers by its unique name.
+	async fn remove_unique(&self, unique_name: &OwnedUniqueName) -> AtspiResult<()> {
+		let mut peers = self.peers.lock().await;
+		peers.retain(|peer| peer.unique_name() != unique_name);
+		Ok(())
+	}
+
+	/// Inserts a new `Peer` with a well-known name into the list of peers.
+	async fn insert_well_known(
+		&self,
+		well_known_name: OwnedWellKnownName,
```

Comment:
Is it true that there can be many names for one application's accessibility interface?